admin user account management

emilia-friedberg opened the pull request:
## WHAT
Update the admin user account creation and management UIs and update the backend to handle all of the different branching paths and emails.

## WHY
We want to allow admins to create accounts for other admins, and better handle all the different possibilities of existing and new user accounts within that creation flow. We also want to make user management more straightforward within this interface.

## HOW
Minor UI updates, more extensive revisions to the backend controller action which sends off the email workers. Note that I moved this method from the `teachers` controller to the `admins` controller because the route was already nested as `admins/:id/teachers`, so it felt confusing to find it in just the `teachers_controller` file, and it isn't really a straightforward create method at all. There are a lot of nested conditionals in there - very open to suggestions about how else we might approach that. Also note that this PR contains both management work and creation work, which was originally a separate branch (#10040).

### Screenshots
<img width="471" alt="Screen Shot 2022-12-07 at 2 56 47 PM" src="https://user-images.githubusercontent.com/18669014/206283395-5fdbb401-6be8-42d2-9d5f-e97fa9814ade.png">
<img width="1023" alt="Screen Shot 2022-12-14 at 10 02 03 AM" src="https://user-images.githubusercontent.com/18669014/207631383-d205e406-0938-4aa3-b18e-1853865b8029.png">

### Notion Card Links
https://www.notion.so/quill/Admin-user-account-creation-and-linking-by-admins-via-admin-dashboard-133bd668e80e450081d37ef0aa630078
https://www.notion.so/quill/Admin-user-account-management-by-admins-via-admin-dashboard-0e57d2ee9c8a432b8006131adb77d472

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? | YES
Have you deployed to Staging? | YES
Self-Review: Have you done an initial self-review of the code below on Github? | YES
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | YES

emilia-friedberg committed:
wip - update admin teacher creation view and messages

emilia-friedberg committed:
merge develop

emilia-friedberg committed:
wip

emilia-friedberg committed:
Merge branch 'develop' of https://github.com/empirical-org/Empirical-Core into feat/admin-user-account-creation

emilia-friedberg committed:
backend and tests

emilia-friedberg committed:
ready to test on staging

emilia-friedberg committed:
update default from and finish set up link

emilia-friedberg committed:
pass back correct value from role

emilia-friedberg committed:
include the "not listed" school under the no school banner

emilia-friedberg committed:
fix lint hopefully

emilia-friedberg committed:
delete leftover code

emilia-friedberg committed:
better handle loading state

emilia-friedberg committed:
wip - backend

emilia-friedberg committed:
wip - getting closer to actually testing

emilia-friedberg committed:
Merge branch 'develop' of https://github.com/empirical-org/Empirical-Core into feat/admin-user-account-creation

emilia-friedberg committed:
Merge branch 'feat/admin-user-account-creation' into feat/admin-user-account-management

emilia-friedberg committed:
things seem to be working but data refresh is a little weird

emilia-friedberg committed:
fix routes

emilia-friedberg committed:
fix translation link and don't send extraneous email

emilia-friedberg committed:
and make sure last sign in data gets populated

emilia-friedberg committed:
one more time with feeling

emilia-friedberg committed:
fix links for admin dashboard in emails

emilia-friedberg committed:
test fix and eric revision

emilia-friedberg committed:
fix lint and test

emilia-friedberg committed:
fix rubocop

emilia-friedberg committed:
tests

emilia-friedberg committed:
remove trailing whitespace

emilia-friedberg committed:
fix another test

eadams17 reviewed (approved):
This looks great! Thanks for getting all of this sorted out 🎉

anathomical commented on services/QuillLMS/app/controllers/admins_controller.rb:
```diff
@@ -43,6 +53,72 @@ def sign_in_account_settings
+  end
+
+  def remove_as_admin
+    SchoolsAdmins.find_by(user_id: params[:id], school_id: params[:school_id]).destroy!
```
Might be worth using a nil-safe accessor on the `find_by` so that we don't call `destroy!` on `nil` if, somehow, the params don't actually correspond to a valid record.

anathomical commented on services/QuillLMS/app/controllers/teachers_controller.rb:
```diff
@@ -5,40 +5,6 @@ class TeachersController < ApplicationController
 
   before_action :require_user, only: [:classrooms_i_teach_with_students, :classrooms_i_own_with_students]
 
-  def create
```
Strong approve of moving the admin-related stuff into the admin controller.

brendanshean commented on services/QuillLMS/spec/serializers/user_admin_serializer_spec.rb:
```diff
@@ -22,22 +23,14 @@
     let!(:school) { create(:school) }
     let!(:schools_admins) { create(:schools_admins, school: school, user: record_instance) }
 
     it 'the #teachers method returns an empty array' do
```
The description should be updated since the array is no longer empty.

brendanshean commented on services/QuillLMS/app/models/user.rb:
```diff
@@ -445,11 +445,11 @@ def satismeter_feature_enabled?
 
-  def admins_teachers
-    schools = administered_schools.includes(:users)
+  def admins_users
```
Maybe rename `admin_user_ids` to better communicate type?

anathomical commented on services/QuillLMS/spec/controllers/admins_controller_spec.rb:
```diff
@@ -44,4 +44,172 @@
+
+  describe '#resend_login_details' do
+    it 'sets the user token and schedules a worker to expire the token in 30 days' do
+      expect(ExpirePasswordTokenWorker).to receive(:perform_in)
```
Maybe not worth doing, but we could be more strict here by adding `.with(30.days, teacher.id)` after the `receive` call to make sure we're getting the delay we expect.

anathomical reviewed (approved):
Looks good!

brendanshean commented on services/QuillLMS/app/serializers/admin/teacher_serializer.rb:
```diff
@@ -4,24 +4,23 @@
+  def schools
+    [object&.school].concat(object&.administered_schools).compact.uniq.map do |school|
+      school_hash = { name: school.name, id: school.id }
+      school_hash[:role] = SchoolsAdmins.find_by(school: school, user: object) ? ADMIN : TEACHER
```
I would use `SchoolsAdmins.exists?(school: school, user: object)` since it is more performant and also communicates that you don't need the object later.

brendanshean commented on services/QuillLMS/app/serializers/user_admin_serializer.rb:
```diff
@@ -1,18 +1,22 @@
   def schools
     admin = User.find(object.id)
-    admin.administered_schools.select("schools.id, schools.nces_id, schools.name")
+    admin.administered_schools.select("schools.id, schools.nces_id, schools.name").sort_by(&:name)
```
what about:  `admin.administered_schools.select(:id, :nces_id, :name).order(:name)`?

brendanshean commented on services/QuillLMS/app/controllers/admins_controller.rb:
```diff
@@ -43,6 +53,72 @@ def sign_in_account_settings
+  def create_and_link_accounts
+    @school = School.find_by(id: params[:school_id])
+
+    if SchoolsAdmins.find_by(school: @school, user: current_user).nil?
```
I would use `exists?` since we're not using the record elsewhere

`unless SchoolsAdmins.exists?(school: @school, user: current_user)`

brendanshean commented on services/QuillLMS/app/controllers/admins_controller.rb:
```diff
@@ -43,6 +53,72 @@ def sign_in_account_settings
+    @school = School.find_by(id: params[:school_id])
+
+    if SchoolsAdmins.find_by(school: @school, user: current_user).nil?
+      render json: {errors: 'Something went wrong. If this problem persists, please contact us at hello@quill.org'}, status: 422
```
I would extract the error message here into `en.yml`

brendanshean commented on services/QuillLMS/app/controllers/admins_controller.rb:
```diff
@@ -43,6 +53,72 @@ def sign_in_account_settings
+        handle_existing_user_submitted_as_teacher
+      end
+    else
+      # Create a new teacher, and automatically join them to the school.
```
I would move this comment into the `handle_new_user` method since it seems to have more functionality now.

brendanshean reviewed (approved):
Some minor comments, but nothing major.  Nice work on pulling all of these different use cases together.  I think the refactoring into admin was a great idea.

emilia-friedberg committed:
remove redirect code from classroom manager

emilia-friedberg committed:
reset form on submission

emilia-friedberg committed:
revisions

emilia-friedberg committed:
fix spec

emilia-friedberg committed:
fix lint error

emilia-friedberg committed:
adjust code to stop checking .nil? on value that should be false

emilia-friedberg committed:
remove tests for old school admin email

eadams17 committed:
Admin user management/ internal tool account creation (#10074)

* convert html erb form to React form

* add logic for adding new and existing users as admins

* fix params

* Try adding a "password" key to the connect string (#10056)

It looks like the old name for this key ("authKey") might be deprecated or something?  At least it didn't work in for me in console.

* handle make admin from cms school edit view

* Staggered Release (#9867)

* get html in place for staggered release modal

* finish teacher ui for staggered release

* standardize tooltip

* update student frontend for staggered release

* finish frontend and add tests

* revisions

* fix test

* Staggered Release Endpoints Part 1 (#9823)

* Rename activity_pack_sequences to pack_sequences

* Rename activity_pack_sequence_activity_packs to pack_sequence_items

* Add staggered release endpoints on backend

* Refactor out bulk save in UnitActivitiesSaver and ClassroomUnitsSaver

* wip

* wip

* ActivityPackSequence getter

* Independent practice recommendations assigner

* wip

* wip

* Remove Units concern

* Add transaction block to bulk create

* Remove ActivityPackSequenceGetter

* Update changes from rename of ActivityPackSequence[ActivityPack] model/table

* Fix some specs

* Staggered Release Endpoints Part 2 (#9839)

* Rename activity_pack_sequences to pack_sequences

* Rename activity_pack_sequence_activity_packs to pack_sequence_items

* add pack_sequence_item_status_mapper

* add pack_sequence_item_status_merger

* Rename merger to combiner

* Staggered Release Endpoints part 3 (#9866)

* Create UserPackSequenceItem model

* Prevent nullable user and pack_sequence_item

* Staggered Release Endpoints Part 4 (#9936)

* Staggered Release Endpoints part 5 (#9967)

* SaveUserPackSequenceItems after changes to activity session state or visible attributes

* Rename active_activity_session spec file

* SaveUserPackSequenceItems after destroying classroom_unit

* SaveUserPackSequenceItems after changes to ClassroomUnit assigned_student_ids or visible attributes

* Add save navigation to classroom

* Temporarily remove unique job

* SaveUserPackSequenceItems after changes to Classroom visible attribute

* SaveUserPackSequenceItems after changes to StudentsClassrooms visible attribute

* Refactor hide_all_activity_sessions to use update_all

* SaveUserPackSequenceItems after changes to UnitActivity visible attribute or record destroyed

* SaveUserPackSequenceItems after changes to Unit visible attribute or record destroyed

* Add in annotation to renamed spec file

* Update some specs

* SaveUserPackSequenceItems after changes to PackSequenceItem order attribute

* Rename activity.is_lesson? to activity.lesson?

* Remove after_save callback on unit

* refactoring activity_session model

* Enable sidekiq-unique-job on SaveUserPackSequenceItemsWorker

* Fix sidekiq AssignRecommendationsWorker job params

* Refactoring on unit

* SaveUserPackSequenceItemsWorker when UserPackSequenceItem is destroyed

* Fix specs

* fix typo

* Fix around_save callback

* misc

* Some refactoring with unit, unit_activity and classroom_unit

* Fix failing specs

* Change some after_save to after_update

* Dan suggestions

* Staggered Release Endpoints part 6 (#9999)

* Remove classroom activity rake task and corresponding worker

* Add save_user_pack_sequence_items to update_all and save(validate: false) calls

* Fix some specs

* remove unnecessary worker

* Dan suggestions

* Staggered Release Endpoints Part 7 (#10036)

* Add sidekiq batch callback for assign recommendations workers

* Dan's suggestions

* Staggered Release Endpoints Part 8 (#10039)

* Check activity links for locks

* Update copy

* Changes for merge with frontend

* Staggered Release Endpoints Part 9 (#10049)

* Remove unit from pack_sequence_item

* Add classroom_unit to pack_sequence_items

* Add null constraints to PackSequence

* Add null constraints to pack_sequence_items

* fix specs

* Add callback to manage changes to assigned_students_ids

* Fix spec

* Fix migration

* Fix race condition

* Emilia suggestion

* fix bug with old pusher channel on recommendations page (#10062)

* do not allow edits to due date and publish date for staggered packs (#10053)

* do not show due date and publish date for staggered packs

* add spec

* fix bug with staying subscribed to old pusher channel

* Fix race condition with checkbox callback (#10068)

* add max height to release method modal (#10058)

* add max height to staggered release method modal

* add overflow: scroll property

Co-authored-by: Emilia Friedberg <emiliafriedberg@gmail.com>

* Fix staggered release locks with multiple pack sequences (#10071)

* Include select of pack_sequence

* Fix calculation of locks for multiple pack_sequences

* Bug fix/ fix spacing issue on featured activity page (#10069)

* fix spacing issue on featured activity page

* remove irrelevant snapshot

* move logic to school and admin controllers; tweak messaging

* fix lint error and tests

* add rubocop exception

* minor refactors

* Fix some intermittent spec failures (#10073)

* Fix intermittent spec failure with google/clever integration cache

* Fix teacher classrooms controller intermittent spec failure

* fix all packs view issue (#10082)

* Update Evidence Banner logic to use new teacher info (#10080)

* Add grade level calculation to teacher_info

* Update teacher concern method.

* Remove stray commit.

* Lint.

* Lint

* update view path

* Fix Staggered Release invalid foreign key (#10081)

* remove save navigation since classroom is non null

* delegate save_user_pack_sequence_items to pack_sequence

* Remove callbacks on user_pack_sequence_item due to circularity

* remove empty line

* Add check for pack_sequence after_destroy callback

* refactor result method to reflect dependency

* Remove unnecessary specs

* wip

* Refactor out pack_sequence_getter

* rescue from race condition

* Add test for PackSequence#save_user_pack_sequence_items

* Move status update into a transaction

* remove callbacks for pack_sequence_item

* refactors

* remove unnecessary check

* small refactors

* fix error reference

* Fix stripe bug with nil credit card (#10086)

* Fix stripe bug with nil credit card

* Update namespace on method

* Fix worker args for whole class instruction assignment (#10091)

* Fix worker args for whole class instruction assignment

* add cast to integer for classroom_id

* Use GitHub for deploy scripts (#10090)

* Revert to using Github branches for deploy scripts

* Use `chmod +x` on LessonsServer deploy.sh

* Remove Sqreen gem from Gemfile and remove related config file (#10094)

* Fix finder bug with User namespace (#10096)

* update scorebook query and view to handle locked activities (#10087)

* update scorebook query and view to handle locked activities

* add tests

* strip sr-only text for toggle highlight function (#10089)

* fix and test (#10088)

* Evidence Activity Health Dashboard front end (#10078)

* all changes

* add poor health flag method

* poor_health_flag method added

* fix linter

* Reacttable changes

* snapshots

* styling fixes

* emma fixes

* comments

* Copy edit/ update Evidence FAQ (#10103)

* update FAQ

* additional edits

* fix score calculation (#10109)

* Admin user management/ segment updates (#10092)

* add staff managed admin event tracking

* add segment calls for remaining workers

* various fixes

* add early returns and some tests

* add remaining tests

* worker and test refactors

* fix lint errors

* remove unneeded comment

* fix worker calls

* fix rubocop

* breakout method

* Remove SKIP_ANNOTATE_SET_DEFAULTS toggle (#10115)

* Fix stripe school subscription bug (#10108)

* Add some event_types to ignored list

* Check for nil subscription status

* fix spacing

* Add check for trialing subscription

* Thomas suggestion

* fix tooltip bugs (#10111)

* migration for closed activity packs view (#10112)

* add migration and spec

* add annotations

* add the frozen string literal

* rename "active" attribute to "open"

* fix argument

* remove return

* fix link and tests

* emilia suggestions

Co-authored-by: Anathomical <thomas.robertson@quill.org>
Co-authored-by: brendanshean <brendanshean@gmail.com>
Co-authored-by: Emilia Friedberg <emiliafriedberg@gmail.com>
Co-authored-by: dandrabik <dandrabik@users.noreply.github.com>
Co-authored-by: cissyxyu <57366100+cissyxyu@users.noreply.github.com>

emilia-friedberg committed:
Merge branch 'develop' into feat/admin-user-account-management

emilia-friedberg committed:
update emails
